include/maliput/geometry_base/kd_tree_strategy.h

```diff
@@ -0,0 +1,87 @@
+// BSD 3-Clause License
+//
+// Copyright (c) 2022, Woven Planet.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+//   list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#pragma once
+
+#include <memory>
+#include <optional>
+#include <vector>
+
+#include "maliput/geometry_base/strategy_base.h"
+#include "maliput/math/kd_tree.h"
+
+namespace maliput {
+namespace geometry_base {
+
+/// Implements StrategyBase by reorganizing the lane space into a kd-tree for
+/// achieving significantly more performant queries than BruteForceStrategy.
+/// The kd-tree is built in construction time by sampling the lanes,
+/// therefore the RoadGeometry should be filled with lanes before
+/// the instantiation.
+
+class KDTreeStrategy : public StrategyBase {
+ public:
+  KDTreeStrategy(const api::RoadGeometry* rg, const double sampling_step);
+  ~KDTreeStrategy() override = default;
+
+ private:
+  class MaliputPoint : public maliput::math::Vector3 {
+   public:
+    MALIPUT_DEFAULT_COPY_AND_MOVE_AND_ASSIGN(MaliputPoint)
+    MaliputPoint(const Vector3& xyz) : Vector3(xyz) {}
```

Comment:
The method `KDTree3D::Nearest()` uses the constructor `MaliputPoint(const Vector3& xyz)`, that's why I left that constructor. Basically the method converts the `Vector3` to a `MaliputPoint` and since we don't know the lane_id because that is exactly what we are looking for we need to use that constructor implicitly. take a look at [  const MaliputPoint maliput_point = kd_tree_->Nearest(point);](https://github.com/maliput/maliput/blob/d7c86db5c7a9c47081e32c029958c5ed0dfdf2fb/src/geometry_base/kd_tree_strategy.cc#L90)

Reply:
Thanks a lot for the explanation. Now I can see clearly where it is used.
I think we should add the explicit keyword though and build it inline with braced initialization for example. I dislike very much implicit conversions because you end up making type promotions in misterious ways. I would rather prefer to have that explicitly annotated.
Can you bear with me a bit more with this? Let's add the explicit keyword there and change the line you linked so you build inline a `MaliputPoint`.

Reply:
Sure, but, I didn't understand what you meant by building inline a MaliputPoint. Something like:
```
  const MaliputPoint maliput_point = kd_tree_->nearest_point(MaliputPoint point.xyz());
```

Reply:
I mean:

```
const MaliputPoint maliput_point = kd_tree_->Nearest(point);